lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php

```diff
@@ -182,41 +211,77 @@ public function walkSelectStatement(SelectStatement $AST)
     }
 
     /**
-     * Generates new SQL for Postgresql or Oracle if necessary.
+     * Finds all PathExpressions in an AST's OrderByClause, and ensures that
+     * the referenced fields are present in the SelectClause of the passed AST.
      *
      * @param SelectStatement $AST
+     */
+    private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
+    {
+        // This block dumps the order by clause node using Node::dump().
+        // It then finds all PathExpressions within and captures the
+        // identificationVariable and field name of each.
+        $orderByDump = (string)$AST->orderByClause;
+        $selects = [];
+        if (preg_match_all('/PathExpression\([^\)]+"identificationVariable": \'([^\']*)\'[^\)]+"field": \'([^\']*)\'[^\)]+\)/i', $orderByDump, $matches, PREG_SET_ORDER)) {
+            foreach($matches as $match) {
+                $idVar = $match[1];
+                $field = $match[2];
+                if (!isset($selects[$idVar])) {
+                    $selects[$idVar] = [];
+                }
+                $selects[$idVar][$field] = true;
+            }
+        }
+
+        // Loop the select clause of the AST and exclude items from $select
+        // that are already being selected.
+        foreach ($AST->selectClause->selectExpressions as $selectExpression) {
+            if ($selectExpression instanceof SelectExpression) {
+                $idVar = $selectExpression->expression;
+                if (!is_string($idVar)) {
+                    continue;
+                }
+                $field = $selectExpression->fieldIdentificationVariable;
+                if ($field === null) {
+                    // No need to add this select, as we're already fetching the whole object.
+                    unset($selects[$idVar]);
+                } else {
+                    unset($selects[$idVar][$field]);
+                }
+            }
+        }
+
+        // Add select items which were not excluded to the AST's select clause.
+        foreach ($selects as $idVar => $fields) {
+            $AST->selectClause->selectExpressions[] = new SelectExpression(new PartialObjectExpression($idVar, array_keys($fields)), null, true);
+        }
+    }
+
+    /**
+     * Generates new SQL for statements with an order by clause
+     *
      * @param array           $sqlIdentifier
      * @param string          $innerSql
      * @param string          $sql
+     * @param OrderByClause   $orderByClause
      *
-     * @return void
+     * @return string
      */
-    public function preserveSqlOrdering(SelectStatement $AST, array $sqlIdentifier, $innerSql, $sql)
+    public function preserveSqlOrdering(array $sqlIdentifier, $innerSql, $sql, $orderByClause)
```

Comment:
Hmmm this is public API. Can we really change method signature and return type here?